editor/src/utils/react-performance.ts

```diff
@@ -209,141 +209,188 @@ function failSafeMemoEqualityFunction(componentDisplayName: string, severity: 's
+      let canSaveOldArray = true
 
-      var newArrayToReturn: any[] = []
-      var canSaveOldArray = true
-      if (length != oldValueInner.length) {
+      length = possibleNewValue.length
+      if (length != oldValue.length) {
         canSaveOldArray = false
       }
       for (i = length; i-- !== 0; ) {
         // try to recurse into the array item here and save it if possible
         newArrayToReturn[i] = keepDeepReferenceEqualityInner(
-          oldValueInner[i],
-          possibleNewValueInner[i],
+          oldValue[i],
+          possibleNewValue[i],
           stackSizeInner + 1,
-          newValueStack,
+          valueStackSoFar,
         )
-        if (oldValueInner[i] !== newArrayToReturn[i]) {
+        if (oldValue[i] !== newArrayToReturn[i]) {
           canSaveOldArray = false
         }
       }
       if (canSaveOldArray) {
-        return oldValueInner
+        return oldValue
       } else {
         return newArrayToReturn
       }
     }
 
-    if (arrA != arrB) return possibleNewValueInner
-
-    var dateA = oldValueInner instanceof Date,
-      dateB = possibleNewValueInner instanceof Date
-    if (dateA != dateB) return possibleNewValueInner
-    if (dateA && dateB)
-      return oldValueInner.getTime() == possibleNewValueInner.getTime()
-        ? oldValueInner
-        : possibleNewValueInner
-
-    var regexpA = oldValueInner instanceof RegExp,
-      regexpB = possibleNewValueInner instanceof RegExp
-    if (regexpA != regexpB) return possibleNewValueInner
-    if (regexpA && regexpB)
-      return oldValueInner.toString() == possibleNewValueInner.toString()
-        ? oldValueInner
-        : possibleNewValueInner
-
-    // for objects we do a deep recursion
-    var keys = keyList(possibleNewValueInner)
+    if (oldValue instanceof Map && possibleNewValue instanceof Map) {
+      let canSaveOldMap = true
+      let newMapToReturn: Map<any, any> = new Map()
+
+      if (oldValue.size !== possibleNewValue.size) {
+        canSaveOldMap = false
+      }
+      for (i of possibleNewValue.entries()) {
+        const oldMapValue = oldValue.get(i[0])
+        const newMapValue = keepDeepReferenceEqualityInner(
+          oldMapValue,
+          i[1],
+          stackSizeInner + 1,
+          valueStackSoFar,
+        )
+        newMapToReturn.set(i[0], newMapValue)
+        if (newMapValue !== oldMapValue) {
+          canSaveOldMap = false
+        }
+      }
+      if (canSaveOldMap) {
+        return oldValue
+      } else {
+        return newMapToReturn
+      }
+    }
+
+    if (oldValue instanceof Set && possibleNewValue instanceof Set) {
+      /**
+       * Sets use reference equality to determine if something is in the set or not,
+       * which makes salvaging sub-values very hard. We don't attempt to do that here
+       * */
+
+      if (oldValue.size !== possibleNewValue.size) return possibleNewValue
+      for (i of oldValue.entries()) if (!possibleNewValue.has(i[0])) return possibleNewValue
```

Comment:
That's some pretty nasty inlining!
```suggestion
      for (entry of oldValue.entries()) {
        if (!possibleNewValue.has(entry[0])) {
          return possibleNewValue
        }
      }
```

Reply:
fixed